### DISCUSSION TITLE
Select Filter with Custom FK

### DISCUSSION THREAD

**Discussion by chaseconey** [filamentphp/filament#8397] [Help] [answered]
I am adding a select filter that I think should be pretty standard, but I will admit that I am pretty new to Filament.

I have 2 related models:

SlackUser, which has a PK of `id`, but a unique id of `slack_id`.

I have a relationship from another model that uses this unique key, rather than the primary key.

```php
public function slackUser()
{
    return $this->belongsTo(SlackUser::class, 'sender', 'slack_id');
}
```

And in Filament, I am trying to use this relatioship as a filter to be able to select based on Slack user. Filter code looks like this:

```php
->filters([
    SelectFilter::make('sender')
        ->relationship('slackUser', 'name')
])
```

The names of the users display properly, but the filter doesn't _seem_ to be working properly. The URL that is generated when applying the filter looks correct to me:

```
?tableFilters[sender][value]=ABC123
```

but if I log out the query, it doesn't seem to be using the custom owner key that is provided in the relationship:

```sql
SELECT
    count(*) AS aggregate
FROM
    "messages"
WHERE (EXISTS (
        SELECT
            *
        FROM
            "slack_users"
        WHERE
            "messages"."sender" = "slack_users"."slack_id"
            AND "slack_users"."id" = 'ABC123'));
```

Any ideas why it wouldn't be picking up the custom owner key in this situation?

Thanks!

**Comment by leandrocfe** [comment]
`->attribute('slack_id')` ?

**Reply by chaseconey** [reply]
I did try that as well, and the select box no longer populates if I do that for some reason. It only has the open `All`.

**Reply by leandrocfe** [reply]
Not sure why. Maybe a relationship issue. I could take a look if you share the project on github.
You can also try a [custom filter form](https://filamentphp.com/docs/3.x/tables/filters#custom-filter-forms)

**Reply by chaseconey** [reply]
Here is an example to show the behavior:

https://github.com/chaseconey/filament-filter-issue

Should be able to run the following steps:

- `cp .env.example .env`
- `composer install && npm i`
- `npm run build`
- `touch database/database.sqlite`
- `php artisan migrate --seed`
- Create a user

And navigate to `/admin` and use the filter in messages.

**Reply by leandrocfe** [reply, marked as answer]
```php
->filters([
    SelectFilter::make('sender')
        ->relationship('slackUser', 'name')
        ->query(
            fn (array $data, Builder $query): Builder =>
            $query->when(
                $data['value'],
                fn (Builder $query, $value): Builder => $query->whereSender($data['value'])
            )
        )
])
```

**Reply by chaseconey** [reply]
Nice - that definitely seems to work.

Any idea if this is a bug? I could submit it as such if so, it seems like we have the information to account for it.

**Reply by leandrocfe** [reply]
Not sure but I'll have a look after

**Reply by ajayjoshioutdosolutions** [reply]
> ```
> array $data, Builder $query
> ```

Thanks for saving my time
